src/codegen/encoding/borsh_encoding.rs

```diff
@@ -693,4 +717,584 @@ impl BorshEncoding {
+                    array_var.clone(),
+                    &dims[(dimension + 1)..dims.len()],
+                    indexes,
+                    true,
+                );
+                cfg.add(
+                    vartab,
+                    Instr::Store {
+                        dest: sub_arr,
+                        data: Expression::Variable(Loc::Codegen, new_ty.clone(), allocated_array),
+                    },
+                );
+            }
+        }
+
+        let for_loop = set_array_loop(array_var, dims, dimension, indexes, vartab, cfg);
+        cfg.set_basic_block(for_loop.body_block);
+        if 0 == dimension {
+            let (read_expr, advance) =
+                self.read_from_buffer(buffer, offset_expr, elem_ty, validator, ns, vartab, cfg);
+            let ptr = load_array_item(array_var, dims, indexes);
+
+            cfg.add(
+                vartab,
+                Instr::Store {
+                    dest: ptr,
+                    data: if matches!(read_expr.ty(), Type::Struct(_)) {
+                        // Type::Struct is a pointer to a struct. If we are dealing with a vector
+                        // of structs, we need to dereference the pointer before storing it at a
+                        // given vector index.
+                        Expression::Load(Loc::Codegen, read_expr.ty(), Box::new(read_expr))
+                    } else {
+                        read_expr
+                    },
+                },
+            );
+            cfg.add(
+                vartab,
+                Instr::Set {
+                    loc: Loc::Codegen,
+                    res: offset_var,
+                    expr: Expression::Add(
+                        Loc::Codegen,
+                        Type::Uint(32),
+                        false,
+                        Box::new(advance),
+                        Box::new(offset_expr.clone()),
+                    ),
+                },
+            );
+        } else {
+            self.decode_complex_array(
+                array_var,
+                buffer,
+                offset_var,
+                offset_expr,
+                dimension - 1,
+                elem_ty,
+                dims,
+                validator,
+                ns,
+                vartab,
+                cfg,
+                indexes,
+            );
+        }
+
+        finish_array_loop(&for_loop, vartab, cfg);
+    }
+
+    /// Read a struct from the buffer
+    fn decode_struct(
+        &self,
+        buffer: &Expression,
+        mut offset: Expression,
+        expr_ty: &Type,
+        struct_ty: &StructType,
+        validator: &mut BufferValidator,
+        ns: &Namespace,
+        vartab: &mut Vartable,
+        cfg: &mut ControlFlowGraph,
+    ) -> (Expression, Expression) {
+        let size = if let Some(no_padding_size) = ns.calculate_struct_non_padded_size(struct_ty) {
+            let padded_size = expr_ty.solana_storage_size(ns);
```

Comment:
Why `solana_storage_size()`? This is not going to be stored in storage, this is going to the heap.

Reply:
How can I know if there is padding between the struct members? If there is no padding, I simply memcpy the struct from the buffer.

Reply:
Any thoughts on this? @seanyoung

Reply:
borsh encoding has nothing to do with solana storage, so it should not use that function. Does size_of() not do the trick?

Reply:
In PR #919, `size_of` has been superseded by `memory_size_of`. It simply sums the size of each item of a struct. I don't know from this number if LLVM will add any type of padding between them. The check I do here serves to verify if a struct can be memcpy'ed from the buffer to the memory. Doing so is more efficient than copying each element individually.

Reply:
@seanyoung Can I memcpy every struct?

Reply:
I created a new function that calculates the padded size correctly.